sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PrivateFieldUsedAsLocalVariable.cs

```diff
@@ -36,340 +36,216 @@ namespace SonarAnalyzer.Rules.CSharp
-
-            var builder = new Dictionary<ISymbol, IDictionary<SyntaxNode, ISymbol>>();
-            foreach (var potentialReference in potentialReferences)
+            public FieldAccessCollector(SemanticModel semanticModel, IDictionary<ISymbol, VariableDeclaratorSyntax> privateFields)
             {
-                var referencedSymbol = semanticModel.GetSymbolInfo(potentialReference).Symbol;
-                if (referencedSymbol == null || !privateFields.ContainsKey(referencedSymbol))
-                {
-                    continue;
-                }
-
-                SyntaxNode referenceSyntax = potentialReference;
-                while (referenceSyntax.Parent != null &&
-                       referencedSymbol.Equals(semanticModel.GetSymbolInfo(referenceSyntax.Parent).Symbol))
-                {
-                    referenceSyntax = referenceSyntax.Parent;
-                }
-
-                if (referenceSyntax.Parent != null &&
-                    referenceSyntax.Parent.IsKind(SyntaxKind.ConditionalAccessExpression))
-                {
-                    referenceSyntax = referenceSyntax.Parent;
-                }
-
-
-                var enclosingSymbol = semanticModel.GetEnclosingSymbol(potentialReference.SpanStart);
-                if (!builder.ContainsKey(enclosingSymbol))
-                {
-                    builder.Add(enclosingSymbol, new Dictionary<SyntaxNode, ISymbol>());
-                }
+                this.semanticModel = semanticModel;
+                this.privateFields = privateFields;
 
-                builder[enclosingSymbol].Add(referenceSyntax, referencedSymbol);
+                privateFieldNames = privateFields.Keys.Select(s => s.Name).ToHashSet();
+                reads = new Dictionary<ISymbol, Dictionary<ISymbol, HashSet<StatementSyntax>>>();
+                writes = new Dictionary<ISymbol, Dictionary<ISymbol, HashSet<StatementSyntax>>>();
             }
 
-            return builder;
-        }
-
-        private static void ExcludePrivateFieldsBasedOnReferences(
-            IImmutableDictionary<ISymbol, PrivateField> privateFields,
-            IDictionary<ISymbol, IDictionary<SyntaxNode, ISymbol>> referencesByEnclosingSymbol,
-            ISet<ISymbol> classMethods)
-        {
-            var referencedAtLeastOnceFromClassMethod = new HashSet<ISymbol>();
-
-            foreach (var references in referencesByEnclosingSymbol)
+            public bool IsRemovableField(ISymbol fieldSymbol)
             {
-                if (!classMethods.Contains(references.Key))
-                {
-                    foreach (var reference in references.Value)
-                    {
-                        privateFields[reference.Value].Excluded = true;
-                    }
+                var writesByMethod = writes.GetValueOrDefault(fieldSymbol);
+                var readsByMethod = reads.GetValueOrDefault(fieldSymbol);

```

Comment:
It's fast access but it's un-needed when `writesByMethod` is `null` so let's move it after the check.